muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/transcriptofrecords/rest/TranscriptofRecordsRESTService.java

```diff
@@ -1,71 +1,315 @@
+    }
+
+    if (!sessionController.isLoggedIn()) {
+      return Response.status(Status.FORBIDDEN).entity("Must be logged in").build();
+    }
+
+    SchoolDataIdentifier studentIdentifier = SchoolDataIdentifier.fromId(studentIdentifierString);
+
+    if (studentIdentifier == null) {
+      return Response.status(Status.NOT_FOUND).entity("Student identifier not found").build();
+    }
+
+    if (!Objects.equals(sessionController.getLoggedUser(), studentIdentifier)) {
+      return Response.status(Status.FORBIDDEN).entity("Can only look at own information").build();
+    }
+
+    User student = userController.findUserByIdentifier(studentIdentifier);
+
+    List<Subject> subjects = courseMetaController.listSubjects();
+    List<VopsRESTModel.VopsRow> rows = new ArrayList<>();
+    int numCourses = 0;
+    int numMandatoryCourses = 0;
+    
+    for (Subject subject : subjects) {
+      if (vopsController.subjectAppliesToStudent(student, subject)) {
+        List<VopsRESTModel.VopsItem> items = new ArrayList<>();
+        for (int i=1; i<MAX_COURSE_NUMBER; i++) {
+          List<Workspace> workspaces =
+              workspaceController.listWorkspacesBySubjectIdentifierAndCourseNumber(
+                  subject.getSchoolDataSource(),
+                  subject.getIdentifier(),
+                  i);
+          List<WorkspaceAssessment> workspaceAssessments = new ArrayList<>();
+          if (!workspaces.isEmpty()) {
+            SchoolDataIdentifier educationSubtypeIdentifier = null;
+            boolean workspaceUserExists = false;
+            for (Workspace workspace : workspaces) {
+              WorkspaceEntity workspaceEntity =
+                  workspaceController.findWorkspaceEntity(workspace);
+              WorkspaceUserEntity workspaceUser =
+                  workspaceUserEntityController.findWorkspaceUserByWorkspaceEntityAndUserIdentifier(
+                      workspaceEntity,
+                      studentIdentifier);
+              SchoolDataIdentifier workspaceIdentifier = new SchoolDataIdentifier(workspace.getIdentifier(), workspace.getSchoolDataSource());
+              workspaceAssessments.addAll(gradingController.listWorkspaceAssessments(workspaceIdentifier, studentIdentifier));
+              if (workspaceUser != null) {
+                workspaceUserExists = true;
+              }
+            }
+            for (Workspace workspace : workspaces) {
+              educationSubtypeIdentifier = workspace.getEducationSubtypeIdentifier();
+              if (educationSubtypeIdentifier != null) {
+                break;
+              }
+            }
+            Mandatority mandatority = educationTypeMapping.getMandatority(educationSubtypeIdentifier);
+            CourseCompletionState state = CourseCompletionState.NOT_ENROLLED;
+            if (workspaceUserExists) {
+              state = CourseCompletionState.ENROLLED;
+            }
+            for (WorkspaceAssessment workspaceAssessment : workspaceAssessments) {
```

Comment:
In addition of there being multiple workspace assessments per workspace, as per above comment, there may be multiple workspaces with the same subject and course number (eg. nonstop and "ryhmäkurssi"). We need to consider the situation where the student has failed one of those courses and passed another.